[Experimental] Product Filters: add individual clear button

roykho opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/46296

This PR adds a "clear filter button" block that is used by all the different filter blocks we have except for the active filters block as that one has its own.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

EXPERIMENTAL: TO BE TESTED BY DEVELOPERS ONLY

1. Go to Appearance->Editor->Templates->Product Catalog
2. Replace Products (beta) block with Products Collection block (catalog)
3. Add all the beta filters (Active, Stock, Attributes, Price, Rating).
4. Select some filters and then click on Clear for any given filter set
5. Ensure only the filter set you have clicked clear is actually cleared.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

Add clear filter button block
</details>

roykho committed:
Product Filters: add individual clear button

roykho committed:
Remove unused style.scss

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

roykho committed:
Simplify searching for the tag

roykho committed:
Convert to use button instead of anchor

roykho committed:
Try using iAPI to show/hide clear button

roykho committed:
Use local context instead of global state

roykho committed:
Fix linting error

roykho committed:
Remove unneeded context

thealexandrelara reviewed (changes requested):
Hey Roy, thank you for working on this, you've made great progress! I noticed some things that could be improved:

### ⚠️ Missing E2E tests
I noticed that some E2E tests are missing to cover these new functionalities. Could you inclusive some E2E tests for them?

### ⚠️ The description is different from the design
In the design, the block's description should be: `Allows shoppers to reset this filter.`

| Current | Expected |
|--------|--------|
| ![CleanShot 2024-05-07 at 16 31 55@2x](https://github.com/woocommerce/woocommerce/assets/20469356/f7464680-22c3-4859-9b31-ac247b194e9d) | ![CleanShot 2024-05-07 at 16 30 59@2x](https://github.com/woocommerce/woocommerce/assets/20469356/84a2859e-4c79-49f3-bebf-4ab06f04f712) | 

### ⚠️ On hover, the Clear button in the editor displays an unexpected visual style

https://github.com/woocommerce/woocommerce/assets/20469356/f290a06c-0247-408b-a85b-9bdb36e90420

### ⛔ Not working on the frontend
When I add one of the Product Filter blocks, save the page and visit the Shop page, an error is occurring and does not allow me to test on the frontend. The error does not occur when I delete the Product Filter blocks. The only block working is the Product Filter: Active Filters (Beta).

https://github.com/woocommerce/woocommerce/assets/20469356/90b9c640-47b2-46a8-a91a-7911ad26d951

roykho commented:
@thealexandrelara - Thanks for the review. Strange I am not seeing any of the issues you're seeing. What theme are you using? Also do you perhaps have Gutenberg plugin active?

> I noticed that some E2E tests are missing to cover these new functionalities. Could you inclusive some E2E tests for them?

Do we add tests for experimental blocks?

roykho committed:
Update clear button block description

roykho committed:
Merge branch 'trunk' into product-filters/clear-button

thealexandrelara commented:
> @thealexandrelara - Thanks for the review. Strange I am not seeing any of the issues you're seeing. What theme are you using? Also do you perhaps have Gutenberg plugin active?

I'm using Twenty-Twenty Four. I just checked, and I'm using Gutenberg 18.2.0. If I disable Gutenberg, the error is gone. I believe it would be good to investigate this to check if this is something caused by us. Otherwise, the block will break websites using the Gutenberg plugin.

> Do we add tests for experimental blocks?

Yes, we do! Before, we handled them in separate PRs, but since our last meetup, it has been defined that we should include them in all new PRs.

roykho committed:
Fix frontend PHP error

roykho committed:
Add e2e tests for attributes filter clear button

roykho committed:
Add e2e tests for price filter clear button

roykho committed:
Add e2e tests for rating filter clear button

roykho committed:
Add e2e tests for stock status filter clear button

roykho committed:
Merge branch 'trunk' into product-filters/clear-button

roykho commented:
Hi @thealexandrelara - I have addressed the following based on your comments:

- Fixed the PHP error when Gutenberg is active (to be fair, I was not aware this PR was causing it so I did not initially test it)
- Added e2e tests

In regards to the visual "oddity" on hover, I think happens when the button is set to "fill"? You can see the same behavior if you add a core/button. By default, it should be set to "outline" and you shouldn't get that issue. Remember, we want users to be able to change it and style it the way they want so we don't want to add any custom CSS styling to it as much as possible and let them control it via the UI.

roykho committed:
Merge branch 'trunk' into product-filters/clear-button

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -97,6 +97,19 @@ test.describe( 'Product Filter: Attribute Block', () => {
+		} ) => {
+			await page.goto( defaultBlockPost.link );
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
Instead of using the css locator, it would be better to use one of the recommended built-in locators, like the `page.getByRole`:

```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

In the above example, it will select a button with the "Clear" text. This is also closer to what users will see on the page.

roykho replied:
I will make the change but I am not sure I totally agree because what if you have multiple buttons on the page that are buttons and with the same name 'Clear'?

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -132,9 +145,89 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -132,9 +145,89 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -132,9 +145,89 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page
+				.locator( '.wp-block-woocommerce-product-filter-clear-button' )
+				.locator( 'button' );
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -132,9 +145,89 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -178,5 +271,95 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -178,5 +271,95 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -178,5 +271,95 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			const button = page
+				.locator( '.wp-block-woocommerce-product-filter-clear-button' )
+				.locator( 'button' );
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/price-filter.block_theme.spec.ts:
```diff
@@ -24,6 +24,19 @@ const test = base.extend< {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/price-filter.block_theme.spec.ts:
```diff
@@ -62,6 +75,86 @@ test.describe( 'Product Filter: Price Filter Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/price-filter.block_theme.spec.ts:
```diff
@@ -62,6 +75,86 @@ test.describe( 'Product Filter: Price Filter Block', () => {
+
+			const button = page
+				.locator( '.wp-block-woocommerce-product-filter-clear-button' )
+				.locator( 'button' );
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/rating-filter.block_theme.spec.ts:
```diff
@@ -24,6 +24,72 @@ const test = base.extend< {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/rating-filter.block_theme.spec.ts:
```diff
@@ -24,6 +24,72 @@ const test = base.extend< {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/rating-filter.block_theme.spec.ts:
```diff
@@ -24,6 +24,72 @@ const test = base.extend< {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/rating-filter.block_theme.spec.ts:
```diff
@@ -24,6 +24,72 @@ const test = base.extend< {
+
+			const button = page
+				.locator( '.wp-block-woocommerce-product-filter-clear-button' )
+				.locator( 'button' );
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -40,6 +40,19 @@ const test = base.extend< {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -71,9 +84,80 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -71,9 +84,80 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -71,9 +84,80 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page
+				.locator( '.wp-block-woocommerce-product-filter-clear-button' )
+				.locator( 'button' );
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -71,9 +84,80 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -90,5 +174,79 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -90,5 +174,79 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -90,5 +174,79 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			const button = page
+				.locator( '.wp-block-woocommerce-product-filter-clear-button' )
+				.locator( 'button' );
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts:
```diff
@@ -178,5 +271,95 @@ test.describe( 'Product Filter: Attribute Block', () => {
+
+			await button.click();
+
+			const placeholder = await page
+				.locator( '.wc-interactivity-dropdown__placeholder' )
+				.textContent();
+
+			expect( placeholder ).toEqual( 'Select Color' );
```
This could be rewritten using the built-in locators:

```suggestion
			const placeholder = page.getByText( 'Select Color' );

			await expect( placeholder ).toBeVisible();
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/price-filter.block_theme.spec.ts:
```diff
@@ -62,6 +75,86 @@ test.describe( 'Product Filter: Price Filter Block', () => {
+			await maxPriceInput.fill( String( defaultMaxRange ) );
+			await maxPriceInput.blur();
+
+			const button = page.locator(
+				'.wp-block-woocommerce-product-filter-clear-button'
+			);
```
```suggestion
			const button = page.getByRole( 'button', { name: 'Clear' } );
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/rating-filter.block_theme.spec.ts:
```diff
@@ -24,6 +24,72 @@ const test = base.extend< {
+
+			await button.click();
+
+			await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
```
I would avoid using `setTimeout` to wait for something to occur, because it will most likely cause flakyness. Playwright includes auto-retrying assertions that remove flakiness by waiting until the condition is met ([reference](https://playwright.dev/docs/actionability)).

roykho replied:
Do you have suggestions around this? Currently without it, it will fail.

roykho replied:
I am able to get around not using timeouts by explicitly asserting it. However for [this here](https://github.com/woocommerce/woocommerce/pull/47101/files#diff-524ce0ba17c55755ebb47ef00ad631b056f0739fcdfa1c659f61d6ddfa74c0a2R136) I am not sure what's the best way to say "wait for" `data-wc-context` to be updated/changed.

roykho replied:
Ok I got around this by checking the URL link. Because after clicking "clear", the URL is expected to have no arguments..

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -71,9 +84,80 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			await button.click();
+
+			await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
```
I suggest removing the fixed setTimeout from here as well.

```suggestion
```

thealexandrelara commented on plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:
```diff
@@ -90,5 +174,79 @@ test.describe( 'Product Filter: Stock Status Block', () => {
+
+			await button.click();
+
+			await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
```
Remove from here as well:

```suggestion
```

thealexandrelara reviewed (changes requested):
Thank you, Roy, for addressing all the issues. Your comment regarding the button's styling makes sense. One thing I noticed is that there currently seems to be no space between the 'Filter' heading and the 'Clear' button (it's more noticeable in TT4 when you hover over the button):

![CleanShot 2024-05-10 at 16 13 42@2x](https://github.com/woocommerce/woocommerce/assets/20469356/ed09c250-5a9c-4bb4-a04a-bfa23a3a9be3)

I left some comments on the code most related to the E2E tests, because some of the custom locators can be replaced by built-in ones. I was also using a lot of CSS locators in my E2E tests, but Sam recommended that I try avoiding them due to their being less resilient to changes and also more prone to introducing flakiness.

roykho commented:
@thealexandrelara - so the spacing issue is correct. Because by default the button is outline. And in that style, the spacing between the heading and clear link is what the design is asking for. If we space that out, then when user first add the block, the spacing will be very far apart and not mimic what the design is.

roykho committed:
Address feedback and use built in locators

roykho committed:
Address feedback and replace the use of timeouts

roykho committed:
Merge branch 'trunk' into product-filters/clear-button

thealexandrelara reviewed (approved):
Awesome! Thank you for working on this! Just tested and everything is working as expected, also all the E2E tests are passing, so I'm approving the PR! 🚀
